modules/FvwmRearrange/FvwmRearrange.1
.\" t
.\" @(#)FvwmRearrange.1  11/9/98
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH FvwmRearrange 1 "1 July 2001"
.UC
.SH NAME
FvwmRearrange \- rearrange FVWM windows
.SH SYNOPSIS
FvwmRearrange is spawned by fvwm, so no command line invocation will work.

.SH DESCRIPTION
This module can be called to tile or cascade windows.

When tiling the module attempts to tile windows on the current screen
subject to certain constraints.  Horizontal or vertical tiling is performed
so that each window does not overlap another, and by default each window
is resized to its nearest resize increment (note sometimes some space
might appear between tiled windows -- this is why).

When cascading the module attempts to cascade windows on the current screen
subject to certain constraints.  Layering is performed so consecutive
windows will have their window titles visible underneath the previous.

.SH INVOCATION
FvwmRearrange is best invoked from a menu, pop up or button.  There are a
number of command line options which can be used to constrain the
layering, these are described below.  As an example case, one could
call FvwmRearrange with the following arguments:
.EX
FvwmRearrange -tile -h 10 10 90 90
.EE
or
.EX
FvwmRearrange -cascade \-resize 10 2 80 70
.EE

The first invocation will horizontally tile windows with a bounding box
which starts at 10 by 10 percent into and down the screen and ends at
90 by 90 percent into and down the screen.

The second invocation will cascade windows starting 10 by 2 percent into and
down the screen.  Windows will be constrained to 80 by 70 percent of
the screen dimensions.  Since the \fIresize\fP is also specified,
windows will be resized to the given constrained width and height.

FvwmRearrange can be called as FvwmTile or FvwmCascade. This is equivalent
to providing the -tile or -cascade option. This form is obsolete and
supplied for backwards compatibility only.

Command-line arguments passed to FvwmRearrange are described here.
.IP \-a
Causes \fIall\fP window styles to be affected, even ones with the
WindowListSkip style.
.IP \-cascade
Cascade windows. This argument must be the first on the command line.
This is the default.
.IP \-desk
Causes all windows on the desk to be cascaded/tiled instead of the
current screen only.
.IP \-flatx
Inhibits border width increment. Only used when cascading.
.IP \-flaty
Inhibits border height increment. Only used when cascading.
.IP \-h
Tiles horizontally (default is to tile vertically). Used for tiling only.
.IP "\-incx \fIarg\fP"
Specifies a horizontal increment which is successively added to
cascaded windows.  \fIarg\fP is a percentage of screen width, or pixel
value if a \fIp\fP is suffixed.  Default is zero. Used only for cascading.
.IP "\-incy \fIarg\fP"
Specifies a vertical increment which is successively added to cascaded
windows.  \fIarg\fP is a percentage of screen height, or pixel value
if a \fIp\fP is suffixed.  Default is zero. Used only for cascading.

.IP \-m
Causes maximized windows to also be affected (implied by \-all).
.IP "\-mn \fIarg\fP"
Tiles up to \fIarg\fP windows in tile direction.  If more windows
exist, a new direction row or column is created (in effect, a matrix
is created). Used only when tiling windows.
.IP \-noraise
Inhibits window raising, leaving the depth ordering intact.
.IP \-noresize
Inhibits window resizing, leaving window sizes intact. This is the default
when cascading windows.
.IP \-nostretch
If tiling: inhibits window growth to fit tile.  Windows are shrunk to fit the
tile but not expanded.

If cascading: inhibits window expansion when using the \-resize option.  Windows
will only shrink to fit the maximal width and height (if given).
.IP \-r
Reverses the window sequence.
.IP \-resize
Forces all windows to resize to the constrained width and height (if
given). This is the default when tiling windows.
.IP \-s
Causes sticky windows to also be affected (implied by \-all).
.IP \-t
Causes transient windows to also be affected (implied by \-all).
.IP \-tile
Tile windows. This argument must be the first on the command line.
.IP \-u
Causes untitled windows to also be affected (implied by \-all).

Up to four numbers can be placed on the command line that are not
switches.  The first pair specify an x and y offset to start the first
window (default is 0, 0).
The meaning of the second pair depends on operation mode:

When tiling windows it specifies an absolute coordinate reference
denoting the lower right bounding box for tiling.

When cascading it specifies a maximal width and height for the layered
windows. If an affected window exceeds either this width or height, it
is resized to the maximal width or height.

If any number is suffixed with the letter p, then it is taken to be a
pixel value, otherwise it is interpreted as a screen percentage.
Specifying zero for any parameter is equivalent to not specifying it.

.SH BUGS
It is probably not a good idea to delete windows while windows are
being rearranged.

.SH AUTHORS
Andrew Veliath (original FvwmTile and FvwmCascade modules)
Dominik Vogt (merged FvwmTile and FvwmCascade to FvwmRearrange)


NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.8 (UNRELEASED)

* Bug fixes:

  - Various DESTDIR fixes (especially around the default-config
    Makefile)
  - fvwm-perllib pod2man fixes.
  - FvwmIconMan no longer triggers a warning about bad size hints
    in the fvwm core.
  - VMware windows do not disapper when switching between
    fullscreen and normal state.
  - Fix (de)installation with the configure options
    --program-prefix, --program-suffix and
    --program-transform-name.

-------------------------------------------------------------------
Changes in stable release 2.6.7 (06-Mar-2016)

* New fvwm features:

  - A new default configuration which is available when fvwm
    doesn't detect a configuration file to load.
  - A new conitional command "Desk n" can restrict matching
    windows to a specific desk.
  - A new conditional command "Screen n" to restrict matching
    windows on a given Xinerama screen.
  - New expansion variable "w.screen" to ascertain the Xineram
    screen number a window is on.
  - New command "InfoStoreClear" to remove all items in the
    InfoStore.

* Removed features:

  - The old and unmaintained debian/and rpm/ directories have
    been remmoved; use the maintainers' copies where available.
  - VMS support has been removed.
  - GTK1.x support has been removed.
  - GNOME-specific window hints (pre-EWMH) have been removed.
  - Some fvwm modules have been removed:
      - FvwmDragWell   (no replacement)
      - FvwmGTK        (no replacement)
      - FvwmSave       (no replacement)
      - FvwmSaveDesk   (no replacement)
      - FvwmScroll     (no replacement)
      - FvwmTabs       (no replacement, never worked anyway)
      - FvwmTaskBar    (use FvwmButtons)
      - FvwmTheme      (in core of fvwm as colorsets)
      - FvwmWharf      (use FvwmButtons)
      - FvwmWinList    (use WindowList command)
      - FvwmWindowMenu (use WindowList command)
      - FvwmIconBox    (use the IconBox style instead)

* New module features:

  - FvwmButtons learned a new option "Colorset" to its
    ChangeButton command.

* fvwm-menu-desktop updated:

  - Renamed default menu to XDGMenu and changed the name
    of the FvwmForm to FvwmForm-XDGMenu-Config to not conflict
    with someone already using FvwmMenu.
  - fvwm-menu-desktop will now load defaults from the
    FvwmForm-XDGMenu-Config data file.
  - Improved dynamic menus by regenerating them on-the-fly.
  - Added new options: --regen-cmd, --dynamic, and more.

* Bug fixes:

  - A bug introduced in 2.6.6 could cause applications with
    negative coordinates to be placed at strange positions.  This
    affected for example acroread when switching to fullscreen
    mode.  This has been fixed.
  - FvwmButtons "Silent" option for dynamic ChangeButton commands
    no longer loops infinitely.

Changes in stable release 2.6.6 (15-Mar-2016)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.
  - Fvwm is much more resilient against applications that flood the
    window manager with repeated events.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-menu-desktop is re-written and provides better support of
    the XDG menu specification.
  - Fix fvwm-menu-desktop keyError bug.  Use "others" if no desktop
    environment found.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The option "NoDeskLimitY" option of the GotoPage command did
    not work.
  - Negative coordinates in the "rectangle" option to the Menu
    commend did not work correctly.  This has been fixed.
  - Removes a slight graphics problem whith the ResizeMaximize
    command being invoked from a window button menu.
  - When an attempt to reparent a client window (i.e. decorate
    it) fails, fvwm no longer throws away all events but only the
    events for that window.

modules/FvwmRearrange/FvwmRearrange.c
/* -*-c-*- */
/*
 * FvwmRearrange.c -- fvwm module to arrange windows
 *
 * Copyright (C) 1996, 1997, 1998, 1999 Andrew T. Veliath
 *
 * Version 1.0
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see: <http://www.gnu.org/licenses/>
 *
 * Combined FvwmTile and FvwmCascade to FvwmRearrange module.
 * 9-Nov-1998 Dominik Vogt
 */
#include "config.h"

#include "libs/ftime.h"
#include <ctype.h>
#include <signal.h>
#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>

#ifdef HAVE_SYS_BSDTYPES_H
#include <sys/bsdtypes.h>
#endif

#if HAVE_SYS_SELECT_H
#include <sys/select.h>
#endif

#include <X11/Xlib.h>

#include "fvwm/fvwm.h"
#include "libs/Module.h"
#include "libs/Parse.h"
#include "libs/System.h"
#include "libs/fvwmlib.h"
#include "libs/vpacket.h"

typedef struct window_item
{
	Window		    frame;
	int		    x, y;
	int		    th, bw;
	int		    width, height;
	int		    min_w, min_h;
	int		    max_w, max_h;
	int		    base_w, base_h;
	int		    inc_w, inc_h;
	struct window_item *prev, *next;
} window_item, *window_list;

/* vars */
Display		  *dpy;
int		   dx, dy;
int		   dwidth, dheight;
static ModuleArgs *module;
int		   fd[2];
fd_set_size_t	   fd_width;
window_list	   wins = NULL, wins_tail = NULL;
int		   wins_count = 0;
FILE		  *console;

/* switches */
int ofsx = 0, ofsy = 0;
int maxw = 0, maxh = 0;
int maxx, maxy;
int untitled = 0, transients = 0;
int maximized	 = 0;
int all		 = 0;
int desk	 = 0;
int current_desk = 0;
int reversed = 0, raise_window = 1;
int resize	= 0;
int nostretch	= 0;
int sticky_page = 0;
int sticky_desk = 0;
int flatx = 0, flaty = 0;
int incx = 0, incy = 0;
int horizontal = 0;
int maxnum     = 0;

int do_maximize = 0;
int do_animate	= 0;
int do_ewmhiwa	= 0;
int is_global	= 0;

int FvwmTile	= 0;
int FvwmCascade = 1;

char	       *monitor_name = NULL;
struct monitor *mon;

RETSIGTYPE
DeadPipe(int sig)
{
	exit(0);
	SIGNAL_RETURN;
}

void
insert_window_list(window_list *wl, window_item *i)
{
	if (*wl) {
		if ((i->prev = (*wl)->prev))
			i->prev->next = i;
		i->next	    = *wl;
		(*wl)->prev = i;
	} else
		i->next = i->prev = NULL;
	*wl = i;
}

void
free_window_list(window_list wl)
{
	window_item *q;

	while (wl) {
		q  = wl;
		wl = wl->next;
		free(q);
	}
}

int
is_suitable_window(unsigned long *body)
{
	XWindowAttributes	xwa;
	struct ConfigWinPacket *cfgpacket = (void *)body;
	struct monitor	       *m = monitor_by_output(cfgpacket->monitor_id);
	char		       *m_name;

	if (m == NULL) {
		fprintf(stderr, "FvwmRearrange: MONITOR WAS NULL\n");
		return 0;
	}
	m_name = (char *)m->si->name;

	if ((DO_SKIP_WINDOW_LIST(cfgpacket)) && !all)
		return 0;

	if ((IS_MAXIMIZED(cfgpacket)) && !maximized)
		return 0;

	if ((IS_STICKY_ACROSS_PAGES(cfgpacket)) && !sticky_page)
		return 0;

	if ((IS_STICKY_ACROSS_DESKS(cfgpacket)) && !sticky_desk)
		return 0;

	if (!XGetWindowAttributes(dpy, cfgpacket->w, &xwa))
		return 0;

	if (xwa.map_state != IsViewable)
		return 0;

	if (!(IS_MAPPED(cfgpacket)))
		return 0;

	if (IS_ICONIFIED(cfgpacket))
		return 0;

	if (desk) {
		if (!is_global && (int)cfgpacket->desk != current_desk)
			return 0;
	} else {
		int x = (int)cfgpacket->frame_x, y = (int)cfgpacket->frame_y;
		int w = (int)cfgpacket->frame_width,
		    h = (int)cfgpacket->frame_height;
		if (x >= dx + dwidth || y >= dy + dheight || x + w <= dx
		    || y + h <= dy)
			return 0;
	}

	if (!(HAS_TITLE(cfgpacket)) && !untitled)
		return 0;

	if ((IS_TRANSIENT(cfgpacket)) && !transients)
		return 0;

	if (!is_global && strcmp(mon->si->name, m_name) != 0)
		return 0;

	return 1;
}

int
get_window(void)
{
	FvwmPacket	       *packet;
	struct ConfigWinPacket *cfgpacket;
	int			last = 0;
	fd_set			infds;

	FD_ZERO(&infds);
	FD_SET(fd[1], &infds);
	select(fd_width, SELECT_FD_SET_CAST & infds, 0, 0, NULL);

	if ((packet = ReadFvwmPacket(fd[1])) == NULL)
		DeadPipe(0);
	else {
		cfgpacket = (struct ConfigWinPacket *)packet->body;
		switch (packet->type &= ~M_EXTENDED_MSG) {
		case M_CONFIGURE_WINDOW:
			if (is_suitable_window(packet->body)) {
				window_item *wi = fxmalloc(sizeof(window_item));
				wi->x		= cfgpacket->frame_x;
				wi->y		= cfgpacket->frame_y;
				wi->frame	= cfgpacket->frame;
				wi->th		= cfgpacket->title_height;
				wi->bw		= cfgpacket->border_width;
				wi->width	= cfgpacket->frame_width;
				wi->height	= cfgpacket->frame_height;
				wi->base_w	= cfgpacket->hints_base_width;
				wi->base_h	= cfgpacket->hints_base_height;
				wi->min_w	= cfgpacket->hints_min_width;
				wi->min_h	= cfgpacket->hints_min_height;
				wi->max_w	= cfgpacket->hints_max_width;
				wi->max_h	= cfgpacket->hints_max_height;
				wi->inc_w	= cfgpacket->hints_width_inc;
				wi->inc_h	= cfgpacket->hints_height_inc;
				if (!wins_tail)
					wins_tail = wi;
				insert_window_list(&wins, wi);
				++wins_count;
			}
			last = 1;
			break;

		case M_END_WINDOWLIST:
			break;

		default:
			fprintf(console,
			    "%s: internal inconsistency: unknown message "
			    "0x%08x\n",
			    module->name, (int)packet->type);
			last = 1;
			break;
		}
	}
	return last;
}

int
atopixel(char *s, unsigned long f)
{
	int l = strlen(s);
	if (l < 1)
		return 0;
	if (isalpha(s[l - 1])) {
		char s2[24];
		strcpy(s2, s);
		s2[strlen(s2) - 1] = 0;
		return atoi(s2);
	}
	return (atoi(s) * f) / 100;
}

void
move_resize_raise_window(window_item *wi, int x, int y, int w, int h)
{
	static char msg[78];
	const char *ewmhiwa = do_ewmhiwa ? "ewmhiwa" : "";
	int orig_w = wi->width - 2 * wi->bw;
	int orig_h = wi->height - 2 * wi->bw - wi->th;

	if (x == wi->x && y == wi->y && (!resize ||
	    (w == orig_w && h == orig_h))) {
		/* Window is the same. Do nothing. */
	} else if (resize) {
		const char *function =
		    do_maximize ? "ResizeMoveMaximize" : "ResizeMove";
		snprintf(msg, sizeof(msg), "%s %dp %dp %up %up %s",
		    function, w, h, x, y, ewmhiwa);
		SendText(fd, msg, wi->frame);
	} else {
		const char *function = do_maximize
					   ? "ResizeMoveMaximize keep keep"
				       : do_animate ? "AnimatedMove"
						    : "Move";
		snprintf(
		    msg, sizeof(msg), "%s %up %up %s", function, x, y, ewmhiwa);
		SendText(fd, msg, wi->frame);
	}

	if (raise_window)
		SendText(fd, "Raise", wi->frame);
}

int
get_hint_size(int val, int base, int min, int max, int inc)
{
	if (val < min)
		val = min;
	if (val > max)
		val = max;
	if (inc > 1) {
		val = base + ((val - base) / inc) * inc;
		if (val < min)
			val += inc;
	}

	return val;
}

void
size_hints_adjustment(window_item *wi, int *w, int *h)
{
	/* Apply size hints */
	*w = get_hint_size(*w, wi->base_w, wi->min_w, wi->max_w, wi->inc_w);
	*h = get_hint_size(*h, wi->base_h, wi->min_h, wi->max_h, wi->inc_h);
}

void
tile_windows(void)
{
	int	     cur_x = ofsx, cur_y = ofsy;
	int	     final_w = -1, final_h = -1;
	int	     wdiv, hdiv, i, j, count = 1;
	window_item *w = reversed ? wins_tail : wins;

	if (horizontal) {
		if ((maxnum > 0) && (maxnum < wins_count)) {
			count = wins_count / maxnum;
			if (wins_count % maxnum)
				++count;
			hdiv = (maxy - ofsy + 1) / maxnum;
		} else {
			maxnum = wins_count;
			hdiv   = (maxy - ofsy + 1) / wins_count;
		}
		wdiv = (maxx - ofsx + 1) / count;

		for (i = 0; w && (i < count); ++i) {
			for (j = 0; w && (j < maxnum); ++j) {
				int nw = wdiv - w->bw * 2;
				int nh = hdiv - w->bw * 2 - w->th;
				size_hints_adjustment(w, &nw, &nh);

				if (resize) {
					if (nostretch) {
						if (nw > w->width)
							nw = w->width;
						if (nh > w->height)
							nh = w->height;
					}
					final_w = (nw > 0) ? nw : w->width;
					final_h = (nh > 0) ? nh : w->height;
				}
				move_resize_raise_window(
				    w, cur_x, cur_y, final_w, final_h);

				cur_y += hdiv;
				w = reversed ? w->prev : w->next;
			}
			cur_x += wdiv;
			cur_y = ofsy;
		}
	} else {
		if ((maxnum > 0) && (maxnum < wins_count)) {
			count = wins_count / maxnum;
			if (wins_count % maxnum)
				++count;
			wdiv = (maxx - ofsx + 1) / maxnum;
		} else {
			maxnum = wins_count;
			wdiv   = (maxx - ofsx + 1) / wins_count;
		}
		hdiv = (maxy - ofsy + 1) / count;

		for (i = 0; w && (i < count); ++i) {
			for (j = 0; w && (j < maxnum); ++j) {
				int nw = wdiv - w->bw * 2;
				int nh = hdiv - w->bw * 2 - w->th;
				size_hints_adjustment(w, &nw, &nh);

				if (resize) {
					if (nostretch) {
						if (nw > w->width)
							nw = w->width;
						if (nh > w->height)
							nh = w->height;
					}
					final_w = (nw > 0) ? nw : w->width;
					final_h = (nh > 0) ? nh : w->height;
				}
				move_resize_raise_window(
				    w, cur_x, cur_y, final_w, final_h);

				cur_x += wdiv;
				w = reversed ? w->prev : w->next;
			}
			cur_x = ofsx;
			cur_y += hdiv;
		}
	}
}

void
cascade_windows(void)
{
	int	     cur_x = ofsx, cur_y = ofsy;
	int	     final_w = -1, final_h = -1;
	window_item *w = reversed ? wins_tail : wins;
	while (w) {
		unsigned long nw = 0, nh = 0;
		if (resize) {
			if (nostretch) {
				if (maxw && (w->width > maxw))
					nw = maxw;
				if (maxh && (w->height > maxh))
					nh = maxh;
			} else {
				nw = maxw;
				nh = maxh;
			}
			if (nw || nh) {
				final_w = nw ? nw : w->width;
				final_h = nh ? nh : w->height;
			}
		}
		move_resize_raise_window(w, cur_x, cur_y, final_w, final_h);

		if (!flatx)
			cur_x += w->bw;
		cur_x += incx;
		if (!flaty)
			cur_y += w->bw + w->th;
		cur_y += incy;
		w = reversed ? w->prev : w->next;
	}
}

void
parse_args(int argc, char *argv[], int argi)
{
	int nsargc = 0;
	/* parse args */
	for (; argi < argc; ++argi) {
		if (!strcmp(argv[argi], "-tile")) {
			FvwmTile    = 1;
			FvwmCascade = 0;
			resize	    = 1;
		} else if (!strcmp(argv[argi], "-cascade")) {
			FvwmCascade = 1;
			FvwmTile    = 0;
		} else if (!strcmp(argv[argi], "-u")) {
			untitled = 1;
		} else if (!strcmp(argv[argi], "-t")) {
			transients = 1;
		} else if (!strcmp(argv[argi], "-a")) {
			all = untitled = transients = maximized = 1;
			if (FvwmCascade) {
				sticky_page = 1;
				sticky_desk = 1;
			}
		} else if (!strcmp(argv[argi], "-r")) {
			reversed = 1;
		} else if (!strcmp(argv[argi], "-noraise")) {
			raise_window = 0;
		} else if (!strcmp(argv[argi], "-noresize")) {
			resize = 0;
		} else if (!strcmp(argv[argi], "-nostretch")) {
			nostretch = 1;
		} else if (!strcmp(argv[argi], "-desk")) {
			desk = 1;
		} else if (!strcmp(argv[argi], "-flatx")) {
			flatx = 1;
		} else if (!strcmp(argv[argi], "-flaty")) {
			flaty = 1;
		} else if (!strcmp(argv[argi], "-r")) {
			reversed = 1;
		} else if (!strcmp(argv[argi], "-h")) {
			horizontal = 1;
		} else if (!strcmp(argv[argi], "-m")) {
			maximized = 1;
		} else if (!strcmp(argv[argi], "-s")) {
			sticky_page = 1;
			sticky_desk = 1;
		} else if (!strcmp(argv[argi], "-sp")) {
			sticky_page = 1;
		} else if (!strcmp(argv[argi], "-sd")) {
			sticky_desk = 1;
		} else if (!strcmp(argv[argi], "-mn") && ((argi + 1) < argc)) {
			maxnum = atoi(argv[++argi]);
		} else if (!strcmp(argv[argi], "-resize")) {
			resize = 1;
		} else if (!strcmp(argv[argi], "-nostretch")) {
			nostretch = 1;
		} else if (!strcmp(argv[argi], "-incx")
			   && ((argi + 1) < argc)) {
			incx = ++argi;
		} else if (!strcmp(argv[argi], "-incy")
			   && ((argi + 1) < argc)) {
			incy = ++argi;
		} else if (!strcmp(argv[argi], "-ewmhiwa")) {
			do_ewmhiwa = 1;
		} else if (!strcmp(argv[argi], "-maximize")) {
			do_maximize = 1;
		} else if (!strcmp(argv[argi], "-nomaximize")) {
			do_maximize = 0;
		} else if (!strcmp(argv[argi], "-animate")) {
			do_animate = 1;
		} else if (!strcmp(argv[argi], "-noanimate")) {
			do_animate = 0;
		} else if (!strcmp(argv[argi], "-screen")
			   && ((argi + 1) < argc)) {
			monitor_name = fxstrdup(argv[++argi]);
		} else {
			if (++nsargc > 4) {
				fprintf(console,
				    "%s: ignoring unknown arg %s\n",
				    module->name, argv[argi]);
				continue;
			}
			if (nsargc == 1) {
				ofsx = argi;
			} else if (nsargc == 2) {
				ofsy = argi;
			} else if (nsargc == 3) {
				maxw = argi;
			} else if (nsargc == 4) {
				maxh = argi;
			}
		}
	}
}

void
update_sizes(char *argv[])
{
	if (incx > 0)
		incx = atopixel(argv[incx], dwidth);
	if (incy > 0)
		incy = atopixel(argv[incy], dheight);
	if (ofsx > 0)
		ofsx = atopixel(argv[ofsx], dwidth);
	if (ofsy > 0)
		ofsy = atopixel(argv[ofsy], dheight);
	if (maxw > 0)
		maxw = atopixel(argv[maxw], dwidth);
	if (maxh > 0)
		maxh = atopixel(argv[maxh], dheight);
	maxx = maxw;
	maxy = maxh;
	ofsx += dx;
	ofsy += dy;
	maxx += dx;
	maxy += dy;
}

int
main(int argc, char *argv[])
{
	char  match[128];
	char *config_line;

	console = fopen("/dev/console", "w");
	if (!console)
		console = stderr;

	module = ParseModuleArgs(argc, argv, 0);
	if (module == NULL) {
		fvwm_debug(__func__,
		    "FvwmRearrange: module should be executed by fvwm only\n");
		exit(-1);
	}

	fd[0] = module->to_fvwm;
	fd[1] = module->from_fvwm;

	if (!(dpy = XOpenDisplay(NULL))) {
		fprintf(console, "%s: couldn't open display %s\n", module->name,
		    XDisplayName(NULL));
		exit(-1);
	}
	signal(SIGPIPE, DeadPipe);

	FScreenInit(dpy);
	fd_width = GetFdWidth();

	/* Need to parse args first so we know what monitor to use. */
	parse_args(module->user_argc, module->user_argv, 0);
	if (monitor_name && StrEquals(monitor_name, "g")) {
		is_global  = 1;
		do_ewmhiwa = 1; /* Ignore hints for global monitor. */
	}

	mon = monitor_resolve_name(monitor_name ? monitor_name : "c");
	if (mon == NULL)
		mon = monitor_get_current();

	dx	= mon->si->x;
	dy	= mon->si->y;
	dwidth	= mon->si->w;
	dheight = mon->si->h;

	strcpy(match, "*");
	strcat(match, module->name);
	InitGetConfigLine(fd, match);
	for (GetConfigLine(fd, &config_line); config_line != NULL;
	     GetConfigLine(fd, &config_line)) {
		char *token, *next, *mname;
		int   dummy, bs_top, bs_bottom, bs_left, bs_right;

		token = PeekToken(config_line, &next);
		if (!StrEquals(token, "Monitor"))
			continue;

		config_line = GetNextToken(next, &mname);
		if (!StrEquals(mname, mon->si->name))
			continue;

		sscanf(config_line, "%d %d %d %d %d %d %d %d %d %d %d %d %d %d",
		    &dummy, &dummy, &dummy, &dummy, &dummy, &dummy, &dummy,
		    &current_desk, &dummy, &dummy, &bs_left, &bs_right, &bs_top,
		    &bs_bottom);
		if (!do_ewmhiwa) {
			dx += bs_left;
			dy += bs_top;
			dwidth -= bs_right;
			dheight -= bs_bottom;
		}
	}

	/* Update sizes now that dwidth and dheight are known */
	update_sizes(module->user_argv);

	SetMessageMask(fd, M_CONFIGURE_WINDOW | M_END_WINDOWLIST);

	if (FvwmTile) {
		if (maxx == dx)
			maxx = dx + dwidth;
		if (maxy == dy)
			maxy = dy + dheight;
	}

	SendText(fd, "Send_WindowList", 0);

	/* Tell fvwm we're running */
	SendFinishedStartupNotification(fd);

	while (get_window()) /* */
		;
	if (wins_count) {
		/* We don't want any more messages from fvwm. */
		SetMessageMask(fd, 0);

		if (FvwmCascade)
			cascade_windows();
		else /* FvwmTile */
			tile_windows();

		/* Wait a second to ensure all messages go through. */
		usleep(1000000);
	}
	free_window_list(wins);

	if (console != stderr)
		fclose(console);

	if (monitor_name)
		free(monitor_name);

	return 0;
}
